cni-repair-controller/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics))
+}
+
+async fn process_events(
+    pod_evts: impl Stream<Item = watcher::Event<Pod>>,
+    tx: Sender<ObjectReference>,
+    metrics: Metrics,
+) {
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let deleting = pod.metadata.deletion_timestamp.is_some();
+
+            if terminated && !deleting {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                let object_ref = pod.object_ref(&());
+                // this avoids blocking the event loop
+                match tx.try_send(object_ref) {
+                    Ok(_) => {}
+                    Err(TrySendError::Full(_)) => {
+                        tracing::warn!(%namespace, %name, "Dropped event (channel full)");
+                        metrics.queue_overflow.inc();
+                    }
+                    Err(TrySendError::Closed(_)) => return,
+                }
+            }
+        }
+    }
+}
+
+async fn process_pods(
+    client: Client,
+    controller_pod_name: String,
+    mut rx: Receiver<ObjectReference>,
+    metrics: Metrics,
+) {
+    while let Some(object_ref) = rx.recv().await {
+        let namespace = object_ref.namespace.clone().unwrap_or_default();
+        let name = object_ref.name.clone().unwrap_or_default();
+        let pods = kube::Api::<Pod>::namespaced(client.clone(), &namespace);
+        let delete_res = pods.delete(&name, &Default::default()).await;
+        match delete_res {
+            Ok(_) => {
+                tracing::info!(%namespace, %name, "Deleting pod");
+                metrics.deleted_pods.inc();
+                if let Err(err) =
+                    publish_k8s_event(client.clone(), controller_pod_name.clone(), object_ref).await
+                {
+                    tracing::warn!(%err, %namespace, %name, "Error publishing event");
+                }
+            }
+            Err(err) => {
+                tracing::warn!(%err, %namespace, %name, "Error deleting pod")
+            }
+        }
```

Comment:
Can either of these async calls block indefinitely? What bounds them?

1. We should apply a timeout to ensure this task can't get "stuck"
2. We should have metrics that count errors for each of these calls
3. We should ideally track a latency histogram for each of these calls.
4. This would be slightly cleaner with an early return :)

```suggestion
        let namespace = object_ref.namespace.clone().unwrap_or_default();
        let name = object_ref.name.clone().unwrap_or_default();
        let pods = kube::Api::<Pod>::namespaced(client.clone(), &namespace);

        let deleted = tokio::select! {
            res = pods.delete(&name, &Default::default() => res,
            _ = time::sleep(time::Duration::from_secs(1)) => {
                todo!("log/count timeout");
                continue;
            }
        };     
        if let Err(err) = deleted {
            tracing::warn!(%err, %namespace, %name, "Error deleting pod");
            todo!("count error");
            continue;
        }
        tracing::info!(%namespace, %name, "Deleted pod");
        metrics.deleted_pods.inc();
        todo!("measure latency");
       
        let event = tokio::select! {
            res = publish_k8s_event(client.clone(), controller_pod_name.clone(), object_ref => res,
            _ = time::sleep(time::Duration::from_secs(1)) => {
                todo!("log/count timeout");
                continue;
            }
        };    
        if let Err(err) = event {
            tracing::warn!(%err, %namespace, %name, "Error publishing event");            
            todo!("count error");
            continue;
        }
        todo!("measure latency");
```

Reply:
Agreed on all accounts. I've pushed an update addressing this and the other comments :+1: